### ISSUE TITLE
Implement Gisting

### ISSUE THREAD

**Issue by abrichr** [OpenAdaptAI/OpenAdapt#27]
How can we take advantage of [gisting](https://github.com/jayelm/gisting)?

[Thread](https://twitter.com/jayelmnop/status/1648365743195684873)

[Paper](https://arxiv.org/abs/2304.08467)

**Comment by FFFiend** [comment]
I've tried mapping the concept to fit Puterbot's case but I'm having trouble understanding what the paper is essentially proposing. Are we compressing prompts into "gist tokens" after initially feeding the LM the entire uncompressed prompt so that it saves on read and compute time later on? Or are we compressing it from the beginning ? The attention masks part is especially confusing, so I was wondering if I could get someone else's perspective on Gisting :D

**Comment by abrichr** [comment]
Thank you for engaging on this issue @FFFiend !

> Are we compressing prompts into "gist tokens" after initially feeding the LM the entire uncompressed prompt so that it saves on read and compute time later on?

I believe that is correct. From https://arxiv.org/pdf/2304.08467.pdf:

> we add a single special gist token to
> the model vocabulary and embedding matrix, much
> like the start/end-of-sentence tokens often present
> in such models. Then, given a (task, input) pair
> (t, x), we concatenate t and x with a set of k successive gist tokens in between: (t, g1, . . . , gk, x), e.g.
> Translate French: <G1> <G2> The cat. This
> sequence is fed into the model, with the restriction that input tokens after the gist tokens cannot
> attend to any of the prompt tokens before the gist
> tokens (but they can attend to the gist tokens). This
> forces the model to compress the information in
> the prompt into the gist tokens, since the input x
> (and output y) cannot attend to the prompt t.

According to the repo (https://github.com/jayelm/gisting) it appears that the trained gist model parameter diffs are now available on HuggingFace for LLaMA-7B and FLAN-T5-XXL 😄

**Comment by FFFiend** [comment]
Alright so I think I spent an unreasonable amount of time trying to understand how exactly the gist tokens themselves looked like from multiple reads on the paper, but I think the repo you linked above blackboxes it for us. 

What I think remains now is to take advantage of the above repo such that any prompting we do on the LLM we use to automate processes, can be gisted and we save on compute time. Unfortunately, I haven't tinkered around with the Puterbot codebase enough yet to try and see where we can inject this improvement :/ Will comment/commit soon :D

**Comment by abrichr** [comment]
@FFFiend please join us on Slack if you haven't already: https://join.slack.com/t/mldsai/shared_invite/zt-1uf94nn7r-qcQnS~hinLPKftUapNzbuw (link at top of README)

**Comment by abrichr** [comment]
> Alright so I think I spent an unreasonable amount of time trying to understand how exactly the gist tokens themselves looked like from multiple reads on the paper,

Thank you!

> but I think the repo you linked above blackboxes it for us.

Unfortunate, but not a deal breaker.

>  I haven't tinkered around with the Puterbot codebase enough yet to try and see where we can inject this improvement

I think the MVP looks like creating a GistingReplayStrategyMixin, analogous to e.g. `LLMReplayStrategyMixin` at https://github.com/MLDSAI/puterbot/blob/main/puterbot/strategies/llm_mixin.py (along the way, perhaps it's worthwhile thinking about what a general framework for any model might look like).

**Comment by abrichr** [comment]
> Edit: Also, I was hoping if you could explain what "diff" meant? I saw it used on the other repo too but I haven't been able to infer what it implies :D

In general a diff is just a difference between two states (e.g. the previous one and the current one). In our case it refers to Screenshots/WindowStates. I'm not sure about what "the other repo" is referencing, can you please clarify? 🙏

**Comment by abrichr** [comment]
> but I think the repo you linked above blackboxes it for us.

Can you please clarify?

From https://github.com/jayelm/gisting#demo--checkpoints:

> To use the model and try out gist caching, use the src/compress.py script, e.g.
> ```
> python -m src.compress --model_name_or_path jayelm/llama-7b-gist-1 --base_llama_path llama-7b \
>     --instruction "Name the top cities in France that should not be missed. Include the best aspects of each place as well."
> ```

**Comment by FFFiend** [comment]
> > Edit: Also, I was hoping if you could explain what "diff" meant? I saw it used on the other repo too but I haven't been able to infer what it implies :D
> 
> In general a diff is just a difference between two states (e.g. the previous one and the current one). In our case it refers to Screenshots/WindowStates. I'm not sure about what "the other repo" is referencing, can you please clarify? 🙏

Oh I meant the Gisting repo used that word too and I was confused what its technical implications were :D


> > but I think the repo you linked above blackboxes it for us.
> 
> Can you please clarify?
> 
> From https://github.com/jayelm/gisting#demo--checkpoints:
> 
> > To use the model and try out gist caching, use the src/compress.py script, e.g.
> > ```
> > python -m src.compress --model_name_or_path jayelm/llama-7b-gist-1 --base_llama_path llama-7b \
> >     --instruction "Name the top cities in France that should not be missed. Include the best aspects of each place as well."
> > ```
> 
> Seems to me like we want this in the `GistingReplayStrategyMixin`, with a `compress` method
Yep noticed this too, apologies on the confusion. The paper didn't go into an example of what the gist tokens looked like and upon a cursory read of the readme it didn't seem like showing the actual compression was something they highlighted, but I was wrong.

**Comment by abrichr** [comment]
@FFFiend if you are interested in being considered for an internship, please submit a PR with your work, and reference this issue in the description.

**Comment by FFFiend** [comment]
Continuing our conversation, I just found out you can make Python modules using an empty __init__ file LOL this task seems a tad bit easier now haha, I was going crazy looking for ways to integrate that ENTIRE codebase to complete the task

**Comment by FFFiend** [comment]
Alright, I sent an email to the owner of the repo (hoping I hear back soon) until then, going from whitepaper to implementation is a little tricky as there are things in files such as compress.py or weight_diff.py that I don't fully understand yet, so any PR at the moment would just be a skeleton ```GistingReplayStrategyMixin``` class with the completion method akin to the ```LLMReplayStrategyMixin``` class, except that the input_token variable would be generated differently, using a new compress method. I think I'm on the right track but I'm not sure ... 


Additionally, I found two possible errors in the ```LLMReplayStrategyMixin``` class. In particular: when we truncate the prompt length to match max_input_size, we have ```prompt = prompt[max_input_size:]``` when it should instead be ```prompt[:max_input_size]``` . Also, the type contract on the max_input_size inside the initializer is off, should be integer. 

I'm holding off making a PR because as stated above, the ```GistingReplayStrategyMixin``` is mostly non-functional until I get a grip on what exactly is going on in the Gisting repo, or if the owner makes it such that I can simply import Gisting as a package somehow. 

Do let me know what you think :D

**Comment by FFFiend** [comment]
Update: A possible solution is to simply include the Gisting repo as a submodule in my fork and work off of that.
